[KEP 4205] Blog post for PSI metrics GA (#54709)
kubernetes/website: content/en/blog/_posts/2026-04-22-psi-metrics-ga/index.md

---
layout: blog
title: "PSI Metrics for Kubernetes Graduates to GA"
date: 2026-XX-XX
draft: true
slug: psi-metrics-ga
author: "Maria Fernanda Romano Silva (Google Cloud)"
---

Since its original implementation in the Linux kernel in 2018, Pressure Stall Information (PSI) has provided users with the high-fidelity signals needed to identify resource saturation before it becomes an outage. Unlike traditional utilization metrics, PSI tells the story of tasks stalled and time lost, all in nicely-packaged percentages of time across the CPU, memory, and I/O.

> Today, we are excited to announce that Kubelet-integrated PSI metrics have graduated to **General Availability (GA)** in Kubernetes v1.36. This graduation ensures that users across the ecosystem have a stable, reliable interface to observe resource contention at the node, pod, and container levels.

Author: Thank you for the callout, I reworded to reference the recent release instead of announcing.

### Beyond Utilization: Why PSI?
Monitoring CPU or memory usage alone can be misleading. A node may report XX% (below 100%) CPU utilization while certain tasks are experiencing severe latency due to scheduling delays. PSI fills this gap by providing:
* **Cumulative Totals**: Absolute time spent in a stalled state.
* **Moving Averages**: 10s, 60s, and 300s windows that allow operators to distinguish between transient spikes and sustained resource tension.

### Proving Stability: Performance Testing at Scale
A common concern when graduating telemetry features is the resource overhead required to collect and serve the metrics. To address this, SIG Node conducted extensive performance validation on high-density workloads (80+ pods) across various machine types.

Our testing focused on four primary scenarios to isolate the impact of the kernel-level trackers and the Kubelet collection logic:
1. **Kernel PSI OFF / Kubelet Feature ON** (Baseline)
2. **Kernel PSI ON / Kubelet Feature ON** (Kernel Scheduler overhead)
3. **Kernel PSI ON / Kubelet Feature OFF** (Default Baseline)

> 4. **Kernel PSI ON / Kubelet Feature ON** (Feature fallback behavior)

Author: I was trying to list out the total conditions but I can see how it was confusing from the overlap. I reworded to clearly outline both cases, one to isolate kernel overhead and the other for kubelet overhead.

{{< figure src="/images/node_sys_cpu_usage_rate_comparison.png" alt="A line graph comparing the Node System (Kernel) CPU usage rate over elapsed time with the PSI feature turned OFF versus ON." title="Node System CPU Usage Rate Comparison" >}}
*Figure 1: Node System CPU comparison under load (80 pods).*

> As seen in Figure 1, the "Kernel Tax" for enabling PSI is remarkably low. Even under heavy I/O and CPU load, the **System CPU** delta between the PSI-enabled (red) and PSI-disabled (blue) clusters remained consistently under **0.2 cores** and over **0.037 cores** for the most part. This confirms that simply enabling the feature does not raise the pre-existing resource use and that the internal kernel bookkeeping for stall tracking is safe for production-scale deployments.

Author: You're right, it was confusing reading back. I added explanations for both cases with their corresponding graphs

{{< figure src="/images/kubelet_cpu_usage_rate_comprison.png" alt="A line graph comparing the Kubelet CPU usage rate over elapsed time with the PSI feature turned OFF versus ON." title="Kubelet CPU Usage Rate Comparison" >}}
*Figure 2: Kubelet CPU Usage Rate Comparison.*

The Kubelet serves as the primary collector for these metrics. Our results show that while the Kubelet performs periodic "sweeps" to aggregate data from the cgroup hierarchy, the CPU usage rate remains stable. The synchronized bursts seen in Figure 2 are virtually identical in both magnitude and frequency, thus attributable to standard Kubelet housekeeping cycles.

You can read more about our performance tests [here](https://docs.google.com/document/d/1ffv7pleid3uk0dT9euK5vkso41i9tscNoEVgeshfiCc/edit?usp=sharing&resourcekey=0-_xKPCU4zqiGU0e7Q8jDJRw).

> ### Getting Started

Reviewer: Do you need to use a particular kernel or does it Just Work?

Author: Yes, in the "Getting Started" section I specified

> ### Getting Started

Author: Great idea! Just added that as well before the "Getting Started" section.

As of v1.36, the `KubeletPSI` feature gate is enabled by default. You can query the Kubelet Summary API to see real-time pressure data:
```

> kubectl get --raw "/api/v1/nodes/$(kubectl get nodes -o jsonpath='{.items[0].metadata.name}')/proxy/stats/summary" | jq '.pods[].containers[] | select(.name=="<CONTAINER_NAME>") | {name, cpu: .cpu.psi, memory: .memory.psi, io: .io.psi}'

Reviewer: Maybe:
```suggestion
CONTAINER_NAME="example-container"
kubectl get --raw "/api/v1/nodes/$(kubectl get nodes -o jsonpath='{.items[0].metadata.name}')/proxy/stats/summary" | jq '.pods[].containers[] | select(.name=="'"$CONTAINER_NAME"'") | {name, cpu: .cpu.psi, memory: .memory.psi, io: .io.psi}'
```

NB, `kubectl -o name` is a thing.

Reviewer: You could also take a Pod name and find the node where it is bound, then proxy to that node.

```

> ### To Learn More

Reviewer: ```suggestion
## Further reading

```

If you want to dive deeper into how these metrics are calculated and exposed, check out these resources:
1. [The official Kernel documentation](https://docs.kernel.org/accounting/psi.html)
2. [Understanding PSI Kubernetes documentation](https://kubernetes.io/docs/reference/instrumentation/understand-psi-metrics/)
3. [cAdvisor Metrics Implementation](https://github.com/google/cadvisor/blob/master/metrics/prometheus.go) 

### Acknowledgements

> This feature was developed through the collaborative efforts of [SIG Node](https://github.com/kubernetes/community/tree/master/sig-node). Special thanks to all contributors who helped design, implement, test, review, and document this feature across its journey from alpha in v1.33, through beta in v1.34, to GA in v1.36.

Reviewer: ```suggestion
Support for PSI metrics was developed through the collaborative efforts of [SIG Node](https://www.kubernetes.dev/community/community-groups/sigs/node/). Special thanks to all contributors who helped design, implement, test, review, and document this feature across its journey from alpha in v1.33, through beta in v1.34, to GA in v1.36.
```

## Beyond utilization: why PSI?

Monitoring CPU or memory usage alone can be misleading. A node may report XX% (below 100%) CPU utilization while certain tasks are experiencing severe latency due to scheduling delays. PSI fills this gap by providing:
* **Cumulative Totals**: Absolute time spent in a stalled state.
* **Moving Averages**: 10s, 60s, and 300s windows that allow operators to distinguish between transient spikes and sustained resource tension.

## Proving stability: performance testing at scale

A common concern when graduating telemetry features is the resource overhead required to collect and serve the metrics. To address this, SIG Node conducted extensive performance validation on high-density workloads (80+ pods) across various machine types.

Our testing focused on two primary scenarios to isolate the impact of the kernel-level and the Kubelet collection respectively:
1. **Kernel PSI OFF / Kubelet Feature ON** vs **Kernel PSI ON / Kubelet Feature ON** (Kernel overhead)

> 2. **Kernel PSI ON / Kubelet Feature OFF** vs **Kernel PSI ON / Kubelet Feature ON** (Kubelet overhead)

Reviewer 2: I still think kubelet overhead is the primary case that we want to measure. Can we reorder these two cases?

Author: Good idea; Applied the reordering.

#### Scenario 1: The Kernel Overhead

> First, we evaluated the underlying overhead of enabling PSI on the Linux kernel. By comparing a cluster booted with `psi=1` against a cluster with `psi=0`, we isolated the exact cost of the OS-level bookkeeping. Even under heavy I/O and CPU load at an 80-pod density, the **System CPU** delta between the kernel-enabled and kernel-disabled clusters remained consistently between **0.037 cores** and **0.2 cores**. This confirms that the internal kernel tracking is highly efficient.

Reviewer 2: > between **0.037 cores** and **0.2 cores**

Do we have data in percentage? The absolute number of cores isn't as comprehensive.

Author: Gotcha, added the percentages now. both scenarios were on 4 core machines so I added that detail as well

> {{< figure src="/images/node_sys_cpu_usage_rate_comparison.png" alt="A line graph comparing the Node System (Kernel) CPU usage rate over elapsed time." title="(Case 1) Node System CPU Usage Rate Comparison" caption="Figure 1: Node system CPU comparison isolating Kubelet overhead under load." >}}

Reviewer 2: The figure is not rendering when I tried to render it on Github. Not sure if it's expected.

Author: At first, I accidentally misspelled "comparson", but the recent commits should've fixed that. Does it render on Netify at least?

Reviewer 2: It does! Thanks!

> {{< figure src="/images/kubeletPSI_sys_cpu_usage_rate_graph.png" alt="A line graph comparing the system CPU usage rate over elapsed time with the PSI feature turned off versus on and kernel PSI off." title="(Case 1) System CPU Usage Rate Comparison" caption="Figure 1: Node System CPU Usage Rate Comparison." >}}

Reviewer 2: You mentioned case 1 is "**Kernel PSI ON / Kubelet Feature OFF** vs **Kernel PSI ON / Kubelet Feature ON**", so I expected kernel PSI to always be `ON` in the comparison. I don't get why the graph is showing `(kernel) PSI ON v.s. OFF`. Did I miss something?

Author: You're correct, Kernal PSI stays ON for case 1. The "PSI On/Off" is referring to the Kubelet feature gate. OH i think I put "off" instead of "on" on the alt field

Author: Just updated the alt field and the graph line labels to be clearer

## Improvements between beta (1.34) and stable (1.36)

- **Smarter Metric Emission for GA:** We improved how the Kubelet handles underlying OS support for PSI. Previously, if the feature was enabled in Kubernetes but the underlying Linux kernel didn't support PSI (`psi=0`), the Kubelet would emit misleading zero-valued metrics. These could trigger false alarms or clutter dashboards when read as real metrics instead of missing values. In v1.36, the Kubelet now detects OS-level PSI support via cgroup configurations before reporting. This ensures that pressure metrics are only collected and emitted when they are actually supported by the node, providing cleaner data for monitoring and alerting systems.

## Getting started

To use PSI metrics in your Kubernetes cluster, your nodes must meet the following requirements:

1. **Ensure your nodes are running a Linux kernel version 4.20 or later and are using cgroup v2.**

> 2. **Ensure PSI is enabled at the OS level** (your kernel must be compiled with `CONFIG_PSI=y` and must not be booted with the `psi=0` parameter).

Reviewer 2: > must not be booted with the `psi=0` parameter

What's the default? If the default is 1, this looks good. If the default it 0, shall we just say must boot with `psi=1` instead?

Author: yup default is 1!
